### DISCUSSION TITLE
Welcome to es_dfm Discussions!

### DISCUSSION THREAD

**Discussion by ThyrixYang** [ThyrixYang/es_dfm#1] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by lichaostoolkit** [comment]
hi, I'm an engineer for recommender system and I have a question about the training of fdp and frn model that if the training data for fdp/frn should contains fake negative sample?  If the data contains fake negative, then the trained fdp model would give a lower probability than real because each delayed positive has a corresponding fake negative. While if the data doesn't contains the fake negative, then how to filter fake negative in streaming protocol? @ThyrixYang

**Reply by ThyrixYang** [reply]
Thank you for your interest in our paper!
We do not need to add fake negatives explicitly in the training data. Please refer to equation 13-16. The p_dp is defined as the probability that a sample is a delayed positive among **all the samples (without fake negatives)**. Thus we simply mark the samples with e < h as positive, and all the others as negative.
The p_rn is defined as the probability that an observed negative sample is a real negative among **all the observed negative samples (with fake negatives)**. Practically, we mark the delayed positives as negative and real negatives as positive, and the positives that are not delayed positive are excluded.
The training of f_dp and f_rn model is on a delayed data stream with ground truth labels(for example, delayed by 7 days), and the labels are constructed as described before.
The code for constructing the pretrianing dataset of ESDFM is at src/data.get_criteo_dataset, tn_dp_pretrain

**Reply by lichaostoolkit** [reply]
Thanks! your answer totally solved my question. I thought that the f_dp and f_rn model use the same data stream with cvr model which is wrong. By the way, if the f_dp and f_rn model use a delayed data stream, would the f_dp/f_rn model be a little out of date?  Because there would be new items and the delayed distribution of a item could be changed.

**Reply by ThyrixYang** [reply]
Yes, the f_dp and f_rn will be different from the ground truth p_dp and p_rn, and this difference will result in a bias which is analyzed in the section "Bias Analysis of Estimated IW". Since the delayed positives are not observed during training, such bias always exists, no matter what method (such as FNW, DFM, FSIW, etc.) is used. The core of delayed feedback problem is to alleviate this bias while improving timeliness.

**Reply by lichaostoolkit** [reply]
yeah, I think esdfm is a more flexible way to solve the delay problem. I'll do some experiment and wish to have some insights :)

**Comment by Maruiruima** [comment]
Hi Yang, I'm trying to figure out the workflow of cvr prediction considering delayed feedback based on your work. If I understand you correctly, the f_rn , f_dp, cvr share the same network structure but with different (or partly different) parameters. The streaming training can be seen as two phases: during the first phase, f_rn and f_dp are trained using separated datasets (noted as pretrain in your code); during the second phase, the parameters in f_cvr is optimized, while the parameters in f_rn and f_dp are fixed, and p_rn and p_dp are inferred from f_rn and f_dp, which are further used in the corrected cvr loss function (eq.17). Then the f_cvr is trained under this loss function, and the value calculated from f_cvr(x') when x' is the test data is seen as the score for ranking used in evaluation. If so, are f_rn and f_dp also incrementally trained as f_cvr? 
I'll appreciate if you could check on the above descriptions, thanks!

**Reply by ThyrixYang** [reply]
Hi, @Maruiruima 
Your descriptions of the workflow of ESDFM are correct.
The f_dp and f_rn models are also incrementally updated but with a larger delay when the dp and rn labels are available. For example, if we label real negatives after a 7 days delay, then we can update f_dp and f_rn after 7 days. We update f_dp and f_rn less frequently (once a day) than f_cvr (once an hour).

**Reply by Maruiruima** [reply]
Thanks for your reply! It helps a lot for me to understand. An additional question: so that is f_cvr trained on the observed q(y|x), while the negative samples contain those that are not converted or not converted immediately (inside observation window), and the positive samples contain both the immediate and the delayed conversion?

**Reply by ThyrixYang** [reply]
Yes, exactly.
